report/src/main/java/com/megaease/easeagent/report/async/zipkin/AgentByteBoundedQueue.java

```diff
@@ -14,126 +14,119 @@
  * the License.
  */
 
-import java.util.Arrays;
-import java.util.concurrent.locks.Condition;
-import java.util.concurrent.locks.ReentrantLock;
+import lombok.Data;
+
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.LongAdder;
 
 /**
  * Multi-producer, multi-consumer queue that is bounded by both count and size.
  *
  * <p>This is similar to {@link java.util.concurrent.ArrayBlockingQueue} in implementation.
  */
 public final class AgentByteBoundedQueue<S> implements WithSizeConsumer<S> {
-    final ReentrantLock lock = new ReentrantLock(false);
-    final Condition available = lock.newCondition();
 
-    final int maxSize;
-    final int maxBytes;
+    private final LinkedTransferQueue<DataWrapper<S>> queue = new LinkedTransferQueue<>();
+
+    private final AtomicInteger sizeInBytes = new AtomicInteger(0);
+
+    private final int maxSize;
 
-    final S[] elements;
-    final int[] sizesInBytes;
-    int count;
-    int sizeInBytes;
-    int writePos;
-    int readPos;
+    private final int maxBytes;
+
+    private final LongAdder loseCounter = new LongAdder();
 
-    @SuppressWarnings("unchecked")
     public AgentByteBoundedQueue(int maxSize, int maxBytes) {
-        this.elements = (S[]) new Object[maxSize];
-        this.sizesInBytes = new int[maxSize];
         this.maxSize = maxSize;
         this.maxBytes = maxBytes;
     }
 
-    /**
-     * Returns true if the element could be added or false if it could not due to its size.
-     */
     @Override
     public boolean offer(S next, int nextSizeInBytes) {
-        lock.lock();
-        try {
-            if (count == maxSize) return false;
-            if (sizeInBytes + nextSizeInBytes > maxBytes) return false;
-
-            elements[writePos] = next;
-            sizesInBytes[writePos++] = nextSizeInBytes;
-
-            if (writePos == maxSize) writePos = 0; // circle back to the front of the array
-
-            count++;
-            sizeInBytes += nextSizeInBytes;
+        if (maxSize == queue.size()) {
+            loseCounter.increment();
+            return false;
+        }
+        if (sizeInBytes.updateAndGet(pre -> pre + nextSizeInBytes) > maxBytes) {
+            loseCounter.increment();
+            sizeInBytes.updateAndGet(pre -> pre - nextSizeInBytes);
+            return false;
+        }
+        queue.offer(new DataWrapper<>(next, nextSizeInBytes));
+        return true;
+    }
 
-            available.signal(); // alert any drainers
-            return true;
-        } finally {
-            lock.unlock();
+    int doDrain(WithSizeConsumer<S> consumer, DataWrapper<S> firstPoll) {
+        int drainedCount = 0;
+        int drainedSizeInBytes = 0;
+        DataWrapper<S> next = null;
+        int count = queue.size() + 1;
```

Comment:
The size method of LinkedTransferQueue requires an O(n) traversal. In this case, we can avoid the call to the size method.

Reply:
Has been resolved.